### PR TITLE
Added diff on vulnerability view

### PR DESCRIPTION
**Pull request by Cheshiriks**
- added history tab
- added changes tab

Снимок экрана 2023-10-05 в 19 18 11

### CODE REVIEW

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            url = "$apiUrl/cosv/cosv-content?cosvFileId=${newCosvFile?.id}",
+            headers = jsonHeaders,
+            loadingHandler = ::loadingHandler,
+        )
+            .unsafeMap {
+                it.text().await()
+            }
+
+        // FixMe: JSON.stringify() don't work in kotlinJs
+        val decodeValue = Json.decodeFromString<RawOsvSchema>(cosvValueNew)
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt:45]: @nulls will be like: 

image

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            url = "$apiUrl/cosv/cosv-content?cosvFileId=${oldCosvFile?.id}",
+            headers = jsonHeaders,
+            loadingHandler = ::loadingHandler,
+        )
+            .unsafeMap {
+                it.text().await()
+            }
+
+        // FixMe: JSON.stringify() don't work in kotlinJs
+        val decodeValue = Json.decodeFromString<RawOsvSchema>(cosvValueOld)
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt:61]: didn't get, why we need decode and encode then if we use only String?

File: save-cosv/src/main/kotlin/com/saveourtool/save/cosv/controllers/CosvController.kt [hunk 3]
```diff
@@ -169,14 +172,34 @@ class CosvController(
...
+    }
+
+    /**
+     * @param cosvFileId
+     * @return cosv file content
+     */
+    @GetMapping("/cosv-content")
+    fun cosvFileContent(
+        @RequestParam cosvFileId: Long,
+    ): Flux<ByteArray> = cosvService.getVulnerabilityVersionAsCosvStream(cosvFileId).map { it.array() }
```
Comment (nulls) [inline save-cosv/src/main/kotlin/com/saveourtool/save/cosv/controllers/CosvController.kt:193]: why you need to call `it.array`?

File: save-cosv/src/main/kotlin/com/saveourtool/save/cosv/controllers/CosvController.kt [hunk 4]
```diff
@@ -169,14 +172,34 @@ class CosvController(
...
 
+    /**
+     * @param identifier
+     * @return list of cosv files
+     */
+    @GetMapping("/list-versions")
+    fun listVersions(
+        @RequestParam identifier: String,
+    ): Flux<CosvFileDto> = blockingToFlux {
+        cosvFileRepository.findAllByIdentifier(identifier).map { it.toDto() }
```
Comment (nulls) [inline save-cosv/src/main/kotlin/com/saveourtool/save/cosv/controllers/CosvController.kt:183]: please do not use `cosvFileRepository` directly, it's internal logic for `cosvStorage`.
Please add this method to `cosvRepository` and add user `cosvStorage` to get list.

Note: `cosvRepository` hides implementation details about storing COSV entries and user `cosvStorage` in current implementation.
`cosvStorage` uses `cosvFileRepository` as a storage for metadata for content saved in S3

**Review by nulls — CHANGES_REQUESTED** [review]

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt (hunk 2 above)
Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt:61]: Ахахахах

Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt:61]: There is a perfect KotlinJs  library called JSON in JS and a wrapper JSON.stringify. 

This method expected to be a JSON formatter, but unfortunately it does not add newlines…

I think we need to investigate it and fix in stdlib, but until it’s done - need to use some existing formatter. 

I suggested the best formatter - prettyPrint in kotlinX. (Probably can be used without decoding, but I have no time to investigate)

Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt:61]: Got it about pretty print, but did you try this one https://stackoverflow.com/questions/4810841/pretty-print-json-using-javascript

Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt:61]: another suggestion -- it's add a special endpoint on backend side which returns StringResponse and return pretty print there

Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt:61]: > Got it about pretty print, but did you try this one https://stackoverflow.com/questions/4810841/pretty-print-json-using-javascript

This is what I have written in above message: "There is a perfect KotlinJs library called JSON in JS and a wrapper JSON.stringify." Need to investigate later, why it does not work.

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/utils/DateTimeUtils.kt [hunk 5]
```diff
@@ -40,6 +40,15 @@ fun LocalDateTime.toUnixCalendarFormat(timeZone: TimeZone = TimeZone.UTC) = toIn
             .minute.plusZero()}"
     }
 
+/**
+ * @param timeZone timezone to print the date time in
+ * @return string representation of [LocalDateTime] in [Thursday, 1 Jan 1970 00:00:00] format
+ */
+fun LocalDateTime.toUnixCalendarFormatWithSeconds(timeZone: TimeZone = TimeZone.UTC) = toInstant(TimeZone.UTC).toLocalDateTime(timeZone)
```
Comment (akuleshov7) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/utils/DateTimeUtils.kt:47]: in future need to merge it with the second method

Comment (akuleshov7) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/utils/DateTimeUtils.kt:47]: with flag

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityHistoryTab.kt [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import com.saveourtool.save.frontend.utils.*
+import com.saveourtool.save.utils.toUnixCalendarFormat
+import react.*
+import react.dom.html.ReactHTML.td
+import web.cssom.ClassName
+
+val vulnerabilityHistoryTab: FC<VulnerabilityHistoryTabProps> = FC { props ->
+
+    val (publicTable) = useState(true)
+    val (t, i18n) = useTranslation(TABLE_HEADERS_LOCALE_NAMESPACE)
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityHistoryTab.kt:17]: you forgot translation

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityHistoryTab.kt [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import com.saveourtool.save.frontend.externals.i18next.useTranslation
+import com.saveourtool.save.frontend.utils.*
+import com.saveourtool.save.utils.toUnixCalendarFormat
+import react.*
+import react.dom.html.ReactHTML.td
+import web.cssom.ClassName
+
+val vulnerabilityHistoryTab: FC<VulnerabilityHistoryTabProps> = FC { props ->
+
+    val (publicTable) = useState(true)
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityHistoryTab.kt:16]: this is useless?

**Comment by akuleshov7** [comment]
return tabs

File: save-cosv/src/main/kotlin/com/saveourtool/save/cosv/controllers/CosvController.kt (hunk 3 above)
Comment (Cheshiriks) [reply save-cosv/src/main/kotlin/com/saveourtool/save/cosv/controllers/CosvController.kt]: Deleted it

File: save-cosv/src/main/kotlin/com/saveourtool/save/cosv/controllers/CosvController.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-cosv/src/main/kotlin/com/saveourtool/save/cosv/controllers/CosvController.kt]: Fixed

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/utils/DateTimeUtils.kt (hunk 5 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/utils/DateTimeUtils.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityHistoryTab.kt (hunk 6 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityHistoryTab.kt:16]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityHistoryTab.kt (hunk 7 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityHistoryTab.kt]: Fixed

**Review by akuleshov7 — APPROVED** [review]

File: save-cosv/src/main/kotlin/com/saveourtool/save/cosv/repository/CosvRepository.kt [hunk 8]
```diff
@@ -74,4 +75,14 @@ interface CosvRepository {
...
         identifier: String,
     ): Flux<LocalDateTime>
+
+    /**
+     * Get all COSV files (versions) with provided [identifier]
+     *
+     * @param identifier
+     * @return [Flux] with all COSV files versions
+     */
+    fun listVersions(
```
Comment (nulls) [inline save-cosv/src/main/kotlin/com/saveourtool/save/cosv/repository/CosvRepository.kt:85]: suggestion to add also:
```kotlin
    /**
     * Downloads COSV from repository as [Flux] of [ByteBuffer] (stream)
     *
     * @param keyId
     * @return [Flux] of [ByteBuffer] with content of COSV
     */
    fun downloadAsStream(
        keyId: Long,
    ): Flux<ByteBuffer>
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt (hunk 2 above)
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/vuln/VulnerabilityChangesTab.kt:61]: let's leave for another PR

File: save-cosv/src/main/kotlin/com/saveourtool/save/cosv/repository/CosvRepository.kt (hunk 8 above)
Comment (nulls) [reply save-cosv/src/main/kotlin/com/saveourtool/save/cosv/repository/CosvRepository.kt:85]: implementation in `CosvRepositoryInStorage`
```kotlin
override fun downloadAsStream(keyId: Long): Flux<ByteBuffer> = cosvFileStorage.downloadByKeyId(keyId)
```
implementation in `CosvFileStorage`:
```kotlin
fun downloadByKeyId(keyId: Long): Flux<ByteBuffer> = blockingToMono { s3KeyManager.findKeyByEntityId(keyId) }
    .switchIfEmptyToNotFound {
        "Not found CosvFile by id $keyId"
    }
    .flatMapMany { download(it) }
```

**Comment by akuleshov7** [comment]
ezgif com-video-to-gif-2

**Pull request merged**